src/utils/Render.js

```diff
@@ -58,52 +30,109 @@ function usingDispatchBuffer(buffer, Component) {
 class DispatchBuffer {
   constructor(renderStrategy) {
     this.promisesBuffer = []
-    this.locked = false
+    this.status = {}
+    this.dispatches = []
     this.renderStrategy = renderStrategy
   }
 
-  push(v) {
-    this.promisesBuffer.push(v)
+  push(id, promise) {
+    this.promisesBuffer.push(promise)
+    promise.then(
+      () => this.status[id] = STAT.DONE,
+      () => this.status[id] = STAT.FAILED
+    )
   }
 
-  fill(Element) {
-    return this.renderStrategy(Element)
+  getStatus(id) {
+    return this.status[id] || STAT.READY
   }
 
   clear() {
     this.promisesBuffer = []
   }
 
-  flush(alt, Element) {
-    return Promise.all(this.promisesBuffer).then((data) => {
-      // fire off all the actions synchronously
-      data.forEach((f) => {
-        if (Array.isArray(f)) {
-          f.forEach(call)
-        } else {
-          call(f)
-        }
-      })
-      this.locked = true
+  resolve(error, html, alt, Element, i) {
+    return Promise.resolve({
+      error,
+      html,
+      state: alt.flush(),
+      buffer: {
+        status: this.status
+      },
+      element: Element,
+      diagnostics: {
+        iterations: i,
+        dispatches: this.dispatches.length
+      }
+    })
+  }
 
-      return {
-        html: this.renderStrategy(Element),
-        state: alt.flush(),
-        element: Element
+  render(alt, Element, info) {
+    alt.recycle()
+
+    const startTime = Date.now()
+    const i = info.i || 0
+
+    // fire off all the actions synchronously
+    this.dispatches.forEach((f) => {
+      if (Array.isArray(f)) {
+        f.forEach(x => x())
+      } else {
+        f()
       }
-    }).catch((err) => {
-      return Promise.reject({
-        err,
-        state: alt.flush(),
-        element: Element
-      })
     })
+
+    // render the html
+    const html = this.renderStrategy(Element)
+
+    if (i >= info.maxIterations) {
+      return this.resolve(
+        new Error('Max number of iterations reached'),
+        html,
+        alt,
+        Element,
+        i
+      )
+    }
+
+    if (info.time > info.timeout) {
+      return this.resolve(
+        new Error('Render timed out'),
+        html,
+        alt,
+        Element,
+        i
+      )
+    }
+
+    // do we have new async queries we need to take care of?
+    if (this.promisesBuffer.length) {
+      // resolve them
+      return Promise.all(this.promisesBuffer).then((data) => {
+        // add the dispatches to our queue
+        this.dispatches = this.dispatches.concat(data)
+
+        // clear the buffer and call render again
+        this.promisesBuffer = []
+
+        info.i = i + 1
+        info.time = info.time + (Date.now() - startTime)
+
+        // this takes the rendering out of a Promise context
+        // TODO test this concurrently
+        return setTimeout(() => this.render(alt, Element, info))
```

Comment:
this isn't necessary actually, im doing this to have a better separation because if a react component throws it'll instead be handled by the Promise which I don't want it to be.